runtime/src/main/java/org/corfudb/runtime/collections/ProtobufIndexer.java

```diff
@@ -38,28 +40,48 @@
         registerSecondaryIndex(payloadSchema, schemaOptions);
     }
 
-    private <T> Index.Spec<Message, CorfuRecord<Message, Message>, ?>
-    getIndex(String indexPath, String indexName, FieldDescriptor fieldDescriptor) {
-        return new Index.Spec<>(
-                () -> indexPath,
-                () -> indexName,
-                (Index.Function<Message, CorfuRecord<Message, Message>, T>)
-                        (key, val) -> ClassUtils.cast(val.getPayload().getField(fieldDescriptor)));
+    static class IndexFieldMapper {
+        private final String[] nestedFields;
+        private final int[] indexMap;
+
+        private final int UNSET = -2;
+
+        public IndexFieldMapper(String indexPath) {
+            // Separate nested fields, as full path is a 'dot' separated String, e.g., 'person.address.street'
+            this.nestedFields = indexPath.split("\\.");
+            this.indexMap = new int[nestedFields.length];
+            Arrays.fill(indexMap, UNSET);
+        }
+
+        public boolean contains(int idx) {
+            return indexMap[idx] != UNSET;
+        }
+
+        public void set(int idx, int value) {
+            indexMap[idx] = value;
+        }
+
+        public int get(int idx) {
+            return indexMap[idx];
+        }
+
+        public String[] getFields() {
+            return nestedFields;
+        }
+
     }
 
     private <T> Index.Spec<Message, CorfuRecord<Message, Message>, ?>
     getNestedIndex(String indexPath, String indexName) {
+        IndexFieldMapper fdMapping = new IndexFieldMapper(indexPath);
```

Comment:
That is the optimization: to have the `IndexFieldMapper` persist it's state across multiple mappings so that it doesn't need to recompute the field numbers via scans every time. The field mappings are lazily computed and apply to Non repeated fields too. 

Note that this doesn't just replace the scan with a direct index look up it also eliminates a lot of string allocations that happen inside of `findFieldByName`